slicerator/__init__.py

```diff
@@ -540,21 +609,31 @@ def _pipeline_fromclass(cls, retain_doc=False):
     retain_doc : bool
         If True, don't modify `func`'s doc string to say that it has been
         made lazy
+    argc : int or 'all', optional
+        Number of arguments that are relevant for the pipeline. Defaults to 1.
 
     Returns
     -------
     Pipeline
         Lazy function evaluation :py:class:`Pipeline` for `func`.
     """
+    if isinstance(argc, str):
+        # subtract 1 for `self`
+        argc = len(inspect.getfullargspec(cls).args) - 1
+
     @wraps(cls)
-    def process(obj, *args, **kwargs):
-        if hasattr(obj, '_slicerator_flag') or isinstance(obj, Slicerator) \
-                or isinstance(obj, Pipeline):
-            return cls(obj, *args, **kwargs)
+    def process(*args, **kwargs):
+        ancestors = args[:argc]
+        args = args[argc:]
+        all_pipe = all(hasattr(a, '_slicerator_flag') or
+                       isinstance(a, Slicerator) or
+                       isinstance(a, Pipeline) for a in ancestors)
+        if all_pipe:
+            return cls(*(ancestors + args), **kwargs)
         else:
             # Fall back on normal behavior of func, interpreting input
             # as a single image.
-            return cls([obj], *args, **kwargs)[0]
+            return cls(*(tuple([a] for a in ancestors) + args), **kwargs)[0]
```

Comment:
This probably doesn't cover the case of 1 indexable and 1 non-indexable ancestor?

Reply:
This is (currently) not supported and also I think it would be non-trivial to implement. A user requiring such a thing could either set `ancestor_count=1` (with reduced flexibility with respect to the second ancestor) or manually turn the second ancestor into a `Slicerator`.

Reply:
I see, this is some kind of array casting. I am :+1: to make the absence of array casting explicit and not allow `ancestor`s with unequal lengths inside a `Pipeline`. So that would include adding a check at the `__init__` and adapting the `__len__`. What do you think?